File: VMSFrontend/src/app/core/customer.models.ts
/**
 * The Customer master and its child records (FSD §10-§15) — the second FSD's own entity, deliberately independent
 * of the Business Partner "Customer" role that already exists in `partner.models.ts` (a different concept that
 * happens to share an English word; see Business Rule §49.1: "Customer is independent from Business Partner").
 */

// ── Customer ───────────────────────────────────────────────────────────────────────

export interface CustomerModel {
  customerId: number;
  customerCode: string;
  customerName: string;
  shortName?: string | null;
  addressLine1: string;
  addressLine2?: string | null;
  countryId: number;
  provinceState?: string | null;
  cityId?: number | null;
  postalCode?: string | null;
  ntn?: string | null;
  strn?: string | null;
  otherRegistrationNo?: string | null;
  currencyCode: string;
  paymentTermsDays: number;
  creditLimit?: number | null;
  status: 'Draft' | 'Active' | 'Inactive';
  inactiveReason?: string | null;
  remarks?: string | null;
  rowVersion: string;
}

export interface SaveCustomerRequest {
  customerCode?: string | null;
  customerName: string;
  shortName?: string | null;
  addressLine1: string;
  addressLine2?: string | null;
  countryId?: number | null;
  provinceState?: string | null;
  cityId?: number | null;
  postalCode?: string | null;
  ntn?: string | null;
  strn?: string | null;
  otherRegistrationNo?: string | null;
  currencyCode?: string | null;
  paymentTermsDays?: number | null;
  creditLimit?: number | null;
  remarks?: string | null;
  rowVersion?: string | null;
}

export interface ChangeCustomerStatusRequest {
  reason?: string | null;
  rowVersion: string;
}

export interface ActivationCheckModel {
  canActivate: boolean;
  missingItems: string[];
}

export interface CustomerHistoryChange {
  id: number;
  occurredAt: string;
  groupId: string;
  userName?: string | null;
  entity: string;
  recordId: string;
  action: string;
  field?: string | null;
  oldValue?: string | null;
  newValue?: string | null;
  reason?: string | null;
  restricted: boolean;
}

export interface CustomerHistory {
  changes: { items: CustomerHistoryChange[]; totalCount: number; page: number; pageSize: number; totalPages: number };
}

export interface CustomerPickerItem {
  customerId: number;
  customerCode: string;
  customerName: string;
}

/** §40A's own per-currency running balance (CC-38) — the List and Details screens (§48.2 screens 1-2) each show
 * it beside the customer, though the full statement/ledger screens themselves are a separate task (§48.5). */
export interface CustomerBalanceModel {
  customerId: number;
  currencyCode: string;
  balanceAmount: number;
}

/** §13A's own currency master — `GET /api/currencies` is gated to `TRP.CURRENCY.MANAGE` (the Currency Setup
 * screen's own Admin-only permission per §48.8), a narrower gate than `TRP.CUSTOMER.EDIT` needs just to populate
 * a currency dropdown; the Customer form degrades to a plain text field when it 403s (documented at the call
 * site) rather than this task widening a permission it does not own. */
export interface CurrencyModel {
  currencyId: number;
  currencyCode: string;
  currencyName: string;
  symbol?: string | null;
  decimalPlaces: number;
  isBase: boolean;
  status: string;
}

// ── Contacts (§11) ─────────────────────────────────────────────────────────────────

export type ContactPurpose = 'Billing' | 'Operations' | 'POD' | 'Other';

export interface CustomerContactModel {
  customerContactId: number;
  customerId: number;
  name: string;
  designation?: string | null;
  mobile1: string;
  mobile2?: string | null;
  telephone?: string | null;
  email?: string | null;
  availabilityTime?: string | null;
  purpose: ContactPurpose[];
  isPrimary: boolean;
  status: 'Active' | 'Inactive';
}

export interface CustomerContactSaveResult {
  contact: CustomerContactModel;
  warnings: string[];
}

export interface SaveCustomerContactRequest {
  name: string;
  designation?: string | null;
  mobile1: string;
  mobile2?: string | null;
  telephone?: string | null;
  email?: string | null;
  availabilityTime?: string | null;
  purpose?: ContactPurpose[] | null;
  isPrimary: boolean;
}

// ── Billing addresses (§12) ────────────────────────────────────────────────────────

export interface CustomerBillingAddressModel {
  customerBillingAddressId: number;
  customerId: number;
  addressName: string;
  addressLine1: string;
  addressLine2?: string | null;
  cityId: number;
  provinceState?: string | null;
  countryId: number;
  postalCode?: string | null;
  ntn?: string | null;
  strn?: string | null;
  isDefault: boolean;
  effectiveFrom: string;
  effectiveTo?: string | null;
  status: 'Active' | 'Inactive';
}

export interface SaveCustomerBillingAddressRequest {
  addressName: string;
  addressLine1: string;
  addressLine2?: string | null;
  cityId: number;
  provinceState?: string | null;
  countryId?: number | null;
  postalCode?: string | null;
  ntn?: string | null;
  strn?: string | null;
  isDefault: boolean;
  effectiveFrom?: string | null;
  effectiveTo?: string | null;
}

// ── Billing configuration (§13) ────────────────────────────────────────────────────

export type DuplicateReferenceBehaviour = 'Allow' | 'Warn' | 'Block';

export interface CustomerBillingConfigurationModel {
  customerBillingConfigurationId: number;
  customerId: number;
  paymentTermsDays: number;
  currencyCode: string;
  defaultBillingAddressId?: number | null;
  defaultInvoiceTemplateId?: number | null;
  invoiceNumberPrefix: string;
  podRequired: boolean;
  evidenceRequired: boolean;
  evidencePageSize: number;
  duplicateReferenceBehaviour: DuplicateReferenceBehaviour;
  customerReferenceRequired: boolean;
  statementEmailContactId?: number | null;
  effectiveFrom: string;
  effectiveTo?: string | null;
}

export interface SaveCustomerBillingConfigurationRequest {
  paymentTermsDays?: number | null;
  currencyCode?: string | null;
  defaultBillingAddressId?: number | null;
  defaultInvoiceTemplateId?: number | null;
  invoiceNumberPrefix?: string | null;
  podRequired: boolean;
  evidenceRequired: boolean;
  evidencePageSize?: number | null;
  duplicateReferenceBehaviour?: DuplicateReferenceBehaviour | null;
  customerReferenceRequired: boolean;
  statementEmailContactId?: number | null;
  effectiveFrom?: string | null;
}

// ── Tax / deduction rules (§14) ────────────────────────────────────────────────────

export type TaxType = 'Percentage' | 'Fixed';
export type CalculationBasis = 'GrossTripAmount' | 'InvoiceSubtotal';

export interface CustomerTaxRuleModel {
  customerTaxRuleId: number;
  customerId: number;
  taxName: string;
  taxCode: string;
  taxType: TaxType;
  taxPercentage?: number | null;
  fixedAmount?: number | null;
  applicable: boolean;
  calculationBasis: CalculationBasis;
  sequence: number;
  effectiveFrom: string;
  effectiveTo?: string | null;
  status: 'Active' | 'Inactive';
  supersedesRuleId?: number | null;
  remarks?: string | null;
}

/** Used for both "add a new rule" (POST) and "replace an existing one" (POST .../replace) — the same body shape. */
export interface SaveCustomerTaxRuleRequest {
  taxName: string;
  taxCode: string;
  taxType: TaxType;
  taxPercentage?: number | null;
  fixedAmount?: number | null;
  applicable: boolean;
  calculationBasis: CalculationBasis;
  sequence: number;
  effectiveFrom?: string | null;
  remarks?: string | null;
  confirmReplace: boolean;
}

/** In-place edit (PUT): identity fields (name/code/type/effective range) are deliberately not here — a rate
 * change goes through {@link SaveCustomerTaxRuleRequest} + `/replace` instead, so it leaves its own history row. */
export interface UpdateCustomerTaxRuleRequest {
  taxPercentage?: number | null;
  fixedAmount?: number | null;
  applicable: boolean;
  calculationBasis?: CalculationBasis | null;
  sequence: number;
  remarks?: string | null;
}

// ── Invoice templates (§15) ────────────────────────────────────────────────────────

export type InvoiceTemplateType = 'SystemStandard' | 'HTMLPDF' | 'DocumentTemplate' | 'ReportDefinition';

export interface CustomerInvoiceTemplateModel {
  customerInvoiceTemplateId: number;
  customerId: number;
  templateName: string;
  version: number;
  templateType: InvoiceTemplateType;
  templateReference: string;
  effectiveFrom: string;
  effectiveTo?: string | null;
  isDefault: boolean;
  status: 'Draft' | 'Active' | 'Inactive';
}

export interface CreateCustomerInvoiceTemplateRequest {
  templateName: string;
  templateType?: InvoiceTemplateType | null;
  templateReference?: string | null;
}

export interface NewInvoiceTemplateVersionRequest {
  templateType?: InvoiceTemplateType | null;
  templateReference?: string | null;
}

export interface ActivateCustomerInvoiceTemplateRequest {
  effectiveFrom?: string | null;
  isDefault: boolean;
}

export interface ApplicableInvoiceTemplatesModel {
  templates: CustomerInvoiceTemplateModel[];
  requiresSelection: boolean;
}

File: src/VMS.Modules.Trips/Models/CustomerModels.cs
namespace VMS.Modules.Trips.Models;

public sealed class CustomerModel
{
    public int CustomerId { get; set; }
    public string CustomerCode { get; set; } = string.Empty;
    public string CustomerName { get; set; } = string.Empty;
    public string? ShortName { get; set; }
    public string AddressLine1 { get; set; } = string.Empty;
    public string? AddressLine2 { get; set; }
    public int CountryId { get; set; }
    public string? ProvinceState { get; set; }
    public int? CityId { get; set; }
    public string? PostalCode { get; set; }
    public string? Ntn { get; set; }
    public string? Strn { get; set; }
    public string? OtherRegistrationNo { get; set; }
    public string CurrencyCode { get; set; } = string.Empty;
    public int PaymentTermsDays { get; set; }
    public decimal? CreditLimit { get; set; }
    public string Status { get; set; } = string.Empty;
    public string? InactiveReason { get; set; }
    public string? Remarks { get; set; }
    public string RowVersion { get; set; } = string.Empty;
}

public sealed class SaveCustomerRequest
{
    /// <summary>Auto-suggested (<c>CUS-00001</c>) when left blank; editable before first use (FSD §10).</summary>
    public string? CustomerCode { get; set; }
    public string CustomerName { get; set; } = string.Empty;
    public string? ShortName { get; set; }
    public string AddressLine1 { get; set; } = string.Empty;
    public string? AddressLine2 { get; set; }
    public int? CountryId { get; set; }
    public string? ProvinceState { get; set; }
    public int? CityId { get; set; }
    public string? PostalCode { get; set; }
    public string? Ntn { get; set; }
    public string? Strn { get; set; }
    public string? OtherRegistrationNo { get; set; }
    public string? CurrencyCode { get; set; }
    public int? PaymentTermsDays { get; set; }
    public decimal? CreditLimit { get; set; }
    public string? Remarks { get; set; }
    /// <summary>Required by every save once the record exists (optimistic concurrency); ignored on create.</summary>
    public string? RowVersion { get; set; }
}

public sealed class ChangeCustomerStatusRequest
{
    /// <summary>Required when setting Inactive; an override reason when force-activating past a failed checklist.</summary>
    public string? Reason { get; set; }
    public string RowVersion { get; set; } = string.Empty;
}

public sealed class ActivationCheckModel
{
    public bool CanActivate { get; set; }
    public IReadOnlyList<string> MissingItems { get; set; } = [];
}

/// <summary>§48.1's own common pattern: "History button on every record opens audit rows (who, when, field, old
/// → new, reason)." Reads the shared, Core-owned `core.AuditEntries` table directly — the same pattern every
/// other module's own history endpoint already hand-rolls (there is no shared history service to call into).
/// No separate "lifecycle" list the way Vehicle has one: a customer's own status changes already show up as
/// ordinary Status field-change rows in <see cref="Changes"/>, so a second table isn't needed to see them.</summary>
public sealed class CustomerHistoryChange
{
    public long Id { get; set; }
    public DateTime OccurredAt { get; set; }
    public Guid GroupId { get; set; }
    public string? UserName { get; set; }
    public string Entity { get; set; } = string.Empty;
    public string RecordId { get; set; } = string.Empty;
    public string Action { get; set; } = string.Empty;
    public string? Field { get; set; }
    public string? OldValue { get; set; }
    public string? NewValue { get; set; }
    public string? Reason { get; set; }
    /// <summary>True when the caller lacks the field permission the change is restricted to (BR-SEC-003) — the
    /// old/new values are withheld, not the row itself, so the change is still visible as having happened.</summary>
    public bool Restricted { get; set; }
}

public sealed class CustomerHistory
{
    public VMS.Shared.Pagination.PaginatedResponse<CustomerHistoryChange> Changes { get; set; } = new();
}
